---
title: "Quality Control"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcbFile: "data/run2.rda"
    minUMIs: 100
    minGenes: 200
    maxGenes: 2500
    maxMitoRatio: 0.1
    minNovelty: 0.70
    outputDir: "."
---

```{r setup, cache=FALSE, message=FALSE, warning=FALSE}
library(bcbioSingleCell)

# Shared RMarkdown settings
prepareSingleCellTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbio object
bcbName <- basename(params$bcbFile) %>%
    # Remove file extension, if specified
    str_replace("\\.[A-Za-z0-9]+", "")
loadDataAsName(c(bcb = params$bcbFile))

# Directory paths
dataDir <- file.path(params$outputDir, "data")
```

```{r header, child="_header.Rmd"}
```



```{r sample_metadata}
sampleMetadata(bcb)
```

[bcbio][] run data was imported from **`r metadata(bcb)[["uploadDir"]]`**.



# Count alignment

We aligned the counts using the latest transcriptome FASTA from FlyBase and obtained additional annotations from [Ensembl][].



# Quality control metrics {.tabset}

## Reads per cell

These are counts of how many reads are assigned to a given cellular barcode. The raw violin and histogram plots on the top show all of the cellular barcodes captured in the experiment. We need to filter out the majority of these noisy, low complexity barcodes from our downstream clustering analysis.

We can better visualize the cellular barcode distribution using a proportional histogram (bottom). Here we should see one single large peak that represents cells that were encapsulated. If we see a strong shoulder, or a bimodal distribution of the cells, that can indicate a couple problems. It might be that there is free floating RNA, which happens when cells are dying. It could also be that there are a set of cells that failed for some reason. Finally, it could also be that there are biologically different types of cells, and one type is much smaller than the other. If this is the case we would expect to see less RNA being sequenced from the smaller cells.

Like the first set of samples analyzed with inDrop, we're seeing a clear bimodal distribution of cellular barcodes. It's likely that the peak near 10^3 reads per cell represents a population of lower complexity cells, like red blood cells. The second peak at 10^4 reads per cell looks good and is fairly typical of inDrop data.

It's important to note that the samples look good with the exception of PC2, which has virtually no cells with a read count above 10^3. There may have been a technical failure for this sample.

```{r plot_reads_per_cell, fig.height=12, fig.width=8}
plotReadsPerCell(bcb)
```


## Cell counts

```{r plot_cell_counts}
plotCellCounts(bcb)
```


## UMI counts per cell

```{r plot_umis_per_cell, fig.height=12, fig.width=8}
plotUMIsPerCell(bcb, min = params$minUMIs)
```


## Genes detected per cell

Here by "detected", we mean genes with a non-zero count measurement per cell. Seeing gene detection in the range of `500`-`5000` is normal for [inDrop][] analysis.

We backed down the gating by genes per cell to a minimum of `r params$minGenes` to allow for inclusion of more cells in the clustering analysis. Additionally, we applied a high end filter of `r params$maxGenes` to filter out what appears to be some noisy barcodes in the UNS OP sample.

```{r plot_genes_per_cell, fig.height=12, fig.width=8}
plotGenesPerCell(bcb, min = params$minGenes, max = params$maxGenes)
```


## UMIs vs. genes detected

If we graph out the total number of UMI counts per cell vs. the genes detected per cell, we can assess whether there is a large population of low quality cells with low counts and/or gene detection.

This should be linear, and looks good for the samples.

```{r plot_umis_vs_genes}
plotUMIsVsGenes(bcb)
```


## Mitochondrial counts ratio

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation.

Normally we recommend applying a max cutoff of 5-10% to filter out stressed or dying cells. There's more of a broad distribution in the samples here than I've seen in other experiments, and apply a 10% cutoff appears to be too strict. Instead, I backed the cutoff down to 30% for the preliminary clustering to allow for more cells. Since the cells were not FACS sorted, it's unlikely that they're highly stressed, and this could actually be bioligical variation we're observing.

I'll be sure to follow up on this in the clustering analysis by coloring the cells by mitochondrial content.

```{r plot_mito_ratio, fig.height=16, fig.width=8}
plotMitoRatio(bcb, max = params$maxMitoRatio)
```


## Novelty

Another way to QC the data is to look for less novelty, that is cells that have less genes detected per count than other cells. We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturated the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that we have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric.

I applied a fairly conservative filtering cutoff here so that we don't discard too many cells prior to clustering.

```{r plot_novelty, fig.height=12, fig.width=8}
plotNovelty(bcb, min = params$minNovelty)
```



# Filter cells

Based on these filtering parameters, we're getting a usable number of cells for clustering, and the counts are fairly close to the expected values. However, it looks like we're seeing a sample failure for PC2, and we likely want to remove that sample from further analysis. We seem to be recovering more cells in the UC sample and fewer in the UNS OP sample than expected.

```{r filter_cells, fig.height=12, fig.width=8, results="asis"}
bcb <- filterCells(
    bcb,
    minUMIs = params$minUMIs,
    minGenes = params$minGenes,
    maxGenes = params$maxGenes,
    maxMitoRatio = params$maxMitoRatio,
    minNovelty = params$minNovelty)
assignAndSaveData(bcbName, bcb)
```



```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```
