---
title: "Quality Control"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcb_file: "data/2018-06-18/bcb.rda"
    min_umis: 100
    max_umis: !r Inf
    min_genes: 100
    max_genes: !r Inf
    max_mito_ratio: 0.25
    min_novelty: 0.85
    min_cells_per_gene: 10
    data_dir: !r file.path("data", Sys.Date())
---

```{r setup, cache=FALSE, message=FALSE}
# Last modified 2018-06-15
# Run this locally using Bioconductor 3.7, since we're using DropletUtils
bcbioSingleCell::prepareSingleCellTemplate()
source("_setup.R")

# Load bcbioSingleCell object
bcb_name <- load(params$bcb_file)
bcb <- get(bcb_name, inherits = FALSE)
stopifnot(is(bcb, "bcbioSingleCell"))
invisible(validObject(bcb))
bcb_original <- bcb

# Temporarily set to `sampleName`
interesting_groups <- interestingGroups(bcb)
interestingGroups(bcb) <- "sampleName"
```

```{r header, child="_header.Rmd"}
```



```{r sample_data}
sampleNames(bcb)
sampleData(bcb, return = "data.frame")
```

[bcbio][] run data was imported from **`r metadata(bcb)$uploadDir`**.



# Transgene expression

Let's double check that the transgenes are expressed

```{r transgenes}
counts(bcb) %>%
    .[c("EGFP", "gal4"), ] %>%
    Matrix::rowSums()
```

The EGFP expression is pretty robust and usable, but the gal4 expression is very low and won't be informative in the downstream analysis.



# Reads per cell {.tabset}

These are counts of how many reads are assigned to a given cellular barcode. It is normal for single cell RNA-seq data to contain a large number of low complexity barcodes. The bcbio pipeline filters out most of these barcodes, and here we have applied a threshold cutoff of a minimum of `r metadata(bcb)$cellularBarcodeCutoff` reads per cell. The unfiltered read count distributions are shown here.


## ECDF

An empirical distribution function (ECDF) plot will show the frequency distribution of the reads per cell. You can see that the vast majority of low complexity barcodes plateau at a read depth below 1000 reads per cell.

```{r plot_reads_per_cell_ecdf}
plotReadsPerCell(bcb, geom = "ecdf")
```


## Histogram

For high quality data, the proportional histogram should contain a single large peak that represents cells that were encapsulated. If we see a strong shoulder, or a bimodal distribution of the cells, that can indicate a couple problems. It might be that there is free floating RNA, which happens when cells are dying. It could also be that there are a set of cells that failed for some reason. Finally, it could also be that there are biologically different types of cells, and one type is much smaller than the other. If this is the case we would expect to see less RNA being sequenced from the smaller cells.

It looks like there a lot of low complexity barcodes that need to be filtered out, but we can see cells with a usable read depth of at least 10,000 (10^4) reads per cell.

```{r plot_reads_per_cell_histogram}
plotReadsPerCell(bcb, geom = "histogram")
```



# UMI counts per cell {.tabset}

Now let's assess the distribution of unique molecular identifier (UMI)-deconvoluted counts per cell. In general, the distributions should be relatively uniform per sample.

This doesn't appear to be the case here. There's a lot of sample-to-sample variation.

Note that uninjured control 1 (from the blood1 pilot) looks to be of lower quality than the samples from blood2 and blood3.


## ECDF

```{r plot_umis_per_cell_ecdf}
# Requires DropletUtils (Bioconductor 3.7)
plotUMIsPerCell(bcb, geom = "ecdf", point = "knee")
```


## Histogram

```{r plot_umis_per_cell_histogram}
plotUMIsPerCell(bcb, geom = "histogram")
```


## Violin

```{r plot_umis_per_cell_violin}
plotUMIsPerCell(bcb, geom = "violin")
```


## Ridgeline

```{r plot_umis_per_cell_ridgeline}
plotUMIsPerCell(bcb, geom = "ridgeline")
```


## Ranks

Now let's rank cellular barcodes based on their UMI count per cell. Note that the scale here is log10. Refer to the `DropletUtils::barcodeRanks()` documentation or the [DropletUtils vignette](https://www.bioconductor.org/packages/3.7/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html) for more information (see "Computing barcode ranks" section).

The knee and inflection points on the curve here help distinguish the UMI threshold between empty droplets with little RNA and cell-containing droplets with much more RNA. The knee point is more sensitive to noise, and we generally recommend starting with the inflection point for filtering because it is more conservative. 

```{r plot_barcode_ranks}
plotBarcodeRanks(bcb)
```



# Filter cells by UMI count

Let's apply this step first and then proceed to evaluating gene detection, mitocondrial transcript abundance, and novelty scores.

```{r filter_cells_by_umi}
bcb <- filterCells(bcb, minUMIs = params$min_umis)
```

Let's take a look at the UMI per cell distributions after this filtering step. Note that we haven't applied very strict filtering here â€” we're going to cut off the "low quality" cells based on the gene detection rate, novelty score, and mitochondrial abundance.

```{r plot_umis_per_cell_postfilter}
plotUMIsPerCell(bcb, geom = "ecdf")
plotUMIsPerCell(bcb, geom = "violin")
```



# UMIs vs. genes detected

If we graph out the total number of UMI counts per cell vs. the genes detected per cell, we can assess whether there is a large population of low quality cells with low counts and/or gene detection.

There's a pretty large pocket of low complexity cellular barcodes that we want to remove prior to clustering. It's a little easier to visualize by calculating the "novelty score" (see below).

```{r plot_umis_vs_genes}
plotUMIsVsGenes(bcb)
```



# Novelty {.tabset}

Another way to QC the data is to look for less novelty, that is cells that have less genes detected per count than other cells. We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturated the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that we have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric.

This metric is very useful for removing "low complexity" cells. This is the primary metric we're using to trash low quality cells.

The injured 2 and uninjured 1 samples are pretty low quality, and most of the cells will be removed. We'll censor these samples from the final dataset.

```{r plot_novelty}
markdownHeader("ECDF", level = 2)
plotNovelty(
    object = bcb,
    geom = "ecdf",
    min = min(params$min_novelty)
)

markdownHeader("Histogram", level = 2)
plotNovelty(
    object = bcb,
    geom = "histogram",
    min = min(params$min_novelty)
)

markdownHeader("Violin", level = 2)
plotNovelty(
    object = bcb,
    geom = "violin",
    min = min(params$min_novelty)
)

markdownHeader("Ridgeline", level = 2)
plotNovelty(
    object = bcb,
    geom = "ridgeline",
    min = min(params$min_novelty)
)
```



# Censor samples

We're dropping injured 2 and uninjured 1 samples due to poor quality. Additionally, we're dropping uninjured 2 because it seems to cluster on its own (see previous QC/clustering reports).

```{r censor_samples}
censor <- c(
    "FACS",
    "sepsis",
    "injured 2",
    "uninjured 1",
    "uninjured 2"
)
```

Let's remove the `r toString(censor)` samples from additional downstream analysis.

```{r select_samples}
print(censor)
keep <- sort(setdiff(sampleNames(bcb), censor))
print(keep)
bcb <- selectSamples(bcb, sampleName = keep)
sampleNames(bcb)
```



# Filter cells by novelty score

```{r filter_cells_by_novelty}
bcb <- filterCells(bcb, minNovelty = params$min_novelty)
```

Now let's take a look at the UMI to gene correlation again after filtering.

```{r plot_after_umi_filtering}
plotUMIsVsGenes(bcb)
plotNovelty(bcb, geom = "ecdf")
plotNovelty(bcb, geom = "violin")
```

See how that's a lot better? We've dropped many of the low complexity cells from the dataset.



# Mitochondrial abundance {.tabset}

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation.

Normally I recommend thresholding at a lower mitochondrial abundance ratio (0.1), but higher expression seems to be biological in *Drosophila*, and is consistent across all of the inDrops runs. So we're thresholding here at a less stringent level instead (`r params$max_mito_ratio`).

```{r plot_mito_ratio}
markdownHeader("ECDF", level = 2)
plotMitoRatio(
    object = bcb,
    geom = "ecdf",
    max = max(params$max_mito_ratio)
)

markdownHeader("Histogram", level = 2)
plotMitoRatio(
    object = bcb,
    geom = "histogram",
    max = max(params$max_mito_ratio)
)

markdownHeader("Violin", level = 2)
plotMitoRatio(
    object = bcb,
    geom = "violin",
    max = max(params$max_mito_ratio)
)

markdownHeader("Ridgeline", level = 2)
plotMitoRatio(
    object = bcb,
    geom = "ridgeline",
    max = max(params$max_mito_ratio)
)
```



# Filter cells by mito abundance

```{r filter_by_mito}
bcb <- filterCells(bcb, maxMitoRatio = params$max_mito_ratio)
```



# Genes detected per cell {.tabset}

Here by "detected", we mean genes with a non-zero count measurement per cell. Seeing gene detection in the range of `500`-`5000` is normal for most single-cell experiments.

The samples look bimodal here in their gene detection distributions, which may indicate a problem with sample preparation. In the meatime, we're going to apply per sample thresholds to return approximately the number of cells expected during the inDrops capture step.

These samples have relatively low average gene per cell counts:

- sepsis
- injured 3
- uninjured 4
- uninjured 5

```{r plot_genes_per_cell}
markdownHeader("ECDF", level = 2)
plotGenesPerCell(
    object = bcb,
    geom = "ecdf",
    min = min(params$min_genes),
    max = max(params$max_genes)
)

markdownHeader("Histogram", level = 2)
plotGenesPerCell(
    object = bcb,
    geom = "histogram",
    min = min(params$min_genes),
    max = max(params$max_genes)
)

markdownHeader("Violin", level = 2)
plotGenesPerCell(
    object = bcb,
    geom = "violin",
    min = min(params$min_genes),
    max = max(params$max_genes)
)

markdownHeader("Ridgeline", level = 2)
plotGenesPerCell(
    object = bcb,
    geom = "ridgeline",
    min = min(params$min_genes),
    max = max(params$max_genes)
)
```



# Filter cells by gene count

Let's threshold all samples except sterile injury 4 by a cutoff of `r params$min_genes` genes per cell. We're thresholding sterile injury 4 at a higher cutoff, to get the number of cells in the ballpark of 3000.

```{r filter_cells_by_gene}
min_genes <- c(
    "injured 1"    = params$min_genes,
    "injured 3"    = params$min_genes,
    "injured 4"    = 650,
    "uninjured 3" = params$min_genes,
    "uninjured 4" = params$min_genes,
    "uninjured 5" = params$min_genes
)
bcb_filtered <- bcb_original %>%
    selectSamples(sampleName = keep) %>%
    filterCells(
        minUMIs = params$min_umis,
        minGenes = min_genes,
        maxGenes = params$max_genes,
        maxMitoRatio = params$max_mito_ratio,
        minNovelty = params$min_novelty,
        minCellsPerGene = params$min_cells_per_gene
    )
```

```{r plot_filtered_qc, results="asis"}
plotQC(
    object = bcb_filtered,
    interestingGroups = "sampleName",
    return = "markdown",
    headerLevel = 2
)
```

```{r save}
saveData(bcb_filtered, dir = params$data_dir)
```



```{r footer, child="_footer.Rmd"}
```
