---
title: "Quality Control"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    minUMIs: 100
    minGenes: 100
    maxGenes: 5000
    maxMitoRatio: 0.1
    minNovelty: 0.75
    outputDir: "."
---

```{r setup, cache=FALSE, message=FALSE, warning=FALSE}
source("setup.R")
load("data/bcb.rda")
```

```{r header, child="_header.Rmd"}
```

[bcbio][] run data was imported from **`r metadata(bcb)[["uploadDir"]]`**.



# FASTQ generation

## inDrop protocol

- `v1`: Original design
    - `_R1`: metadata read
    - `_R2`: biological read
- `v2`: Inversion of `v1`
- `v3`: Summer 2016 redesign, requiring manual demultiplexing
    - `_R1`: biological read
    - `_R2`: first half of the gel barcode
    - `_R3`: library index
    - `_R4`: second half of the gel barcode, the UMI, and a fraction of the polyA tail

This study used the [inDrop][] `v3` design.


## Demultiplexing

We demultiplexed the raw Illumina BCL run files provided by the [Harvard Biopolymers Facility][bpf] into FASTQ format using [bcl2fastq][]. This step generates files with the following suffixes:

- `_R1`: 61 bp read 1, transcript
- `_R2`: 8 bp index read 1 (i7), single cell barcode
- `_R3`: 8 bp index read 2 (i5), library index
- `_R4`: 14 bp read 2, barcode and unique molecular identifiers (UMIs)

We ran this on [Orchestra][] with the following settings:

```{bash bcl2fastq, echo=TRUE, eval=FALSE}
module load seq/bcl2fastq/2.17.1.14
bcl2fastq --use-bases-mask y*,y*,y*,y* --mask-short-adapter-reads 0 --minimum-trimmed-read-length 0
```



# Count alignment

We aligned the counts using the latest transcriptome FASTA from [FlyBase][], [`dmel-all-transcript-r6.14.fasta`](ftp://ftp.flybase.net/genomes/Drosophila_melanogaster/dmel_r6.14_FB2017_01/fasta/dmel-all-transcript-r6.14.fasta.gz) (Feb 2017). We used the matching GTF to obtain transcript-to-gene identifier mappings, and additional gene annotation metadata from [Ensembl][].



# Quality control metrics {.tabset}

## Reads per cell

These are counts of how many reads are assigned to a given cellular barcode. The purpose of these histograms is that we should see one single large peak that represents cells that were encapsulated. If we see a strong shoulder, or a bimodal distribution of the cells, that can indicate a couple problems. It might be that there is free floating RNA, which happens when cells are dying. It could also be that there are a set of cells that failed for some reason. Finally, it could also be that there are biologically different types of cells, and one type is much smaller than the other. If this is the case we would expect to see less RNA being sequenced from the smaller cells.

Both the `sorted` and `unsorted` groups have an apparent bimodal distribution, which could be indicative of a large number of dying cells or low complexity cells, possibly red blood cells. We're not entirely sure what's behind this effect, and we are working on figuring this out with the Harvard Single Cell Core.

```{r plot_reads_per_cell, fig.height=12, fig.width=8}
plotReadsPerCell(bcb)
```


## Total number of cells

More cells were analyzed in the FACS sorted group than the expected amount of 2,000. We obtained a low cell count in the unsorted group.

```{r plot_cell_counts}
plotCellCounts(bcb)
```


## Total RNA read counts

When sequencing more cells, we have overall less counts per cell. The distribution of counts here looks reasonable; we are truncating the lower end because we required there to be at least 10,000 reads to match a barcode before aligning. The count depth for the unsorted group is lower than we'd like to see.

```{r plot_umis_per_cell, fig.height=12, fig.width=8}
plotUMIsPerCell(bcb, min = params$minUMIs)
```


## Genes detected

Here by "detected", we mean genes with a non-zero count measurement per cell. Seeing gene detection in the range of ~2,000-2,500 is normal for [inDrop][] analysis. The low gene detection rate of the unsorted group is likely problematic.

```{r plot_genes_per_cell, fig.height=12, fig.width=8}
plotGenesPerCell(bcb, min = params$minGenes, max = params$maxGenes)
```


## Total vs. detected

If we graph out the total number of counts per cell vs. the genes detected per cell, we can assess whether there is a large population of low quality cells with low counts and/or gene detection. We're not seeing this in the FACS sorted group, which is good. There's more spread in the unsorted group, but not a huge shoulder of low complexity, which has been seen in some other studies.

```{r plot_umis_vs_genes}
plotUMIsVsGenes(bcb)
```


## Mitochondrial counts

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation. We're not seeing very high mitochondrial gene expression (> 20k counts) in most cells, so that's good. The difference looks large between the `sorted` and `unsorted` groups here because the count depth is relatively low in the `unsorted` group. We've seen samples in other studies with higher mitochondrial gene expression.

```{r plot_mito_ratio, fig.height=16, fig.width=8}
plotMitoRatio(bcb, max = params$maxMitoRatio)
```


## Novelty

Another way to QC the data is to look for less novelty, that is cells that have less genes detected per count than other cells. We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturated the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that we have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric.

```{r plot_novelty, fig.height=12, fig.width=8}
plotNovelty(bcb, min = params$minNovelty)
```



# Filter cells

We can continue on to clustering analysis with the FACS sorted sample. The unsorted sample has low cell counts and low gene detection. As mentioned above, the FACS sorted sample has a bimodal distribution of the cellular barcodes, which is not desirable. We'll work on figuring out the reason behind this. We don't see a big shoulder of low complexity cells in the FACS group, and we recommend applying only minimal filtering.

```{r filter_cells, fig.height=12, fig.width=8, results="asis"}
bcb_filtered <- filterCells(
    bcb,
    minUMIs = params$minUMIs,
    minGenes = params$minGenes,
    maxGenes = params$maxGenes,
    maxMitoRatio = params$maxMitoRatio,
    minNovelty = params$minNovelty)
saveData(bcb_filtered)
```



```{r footer, child="_footer.Rmd"}
```
